src/Functions/parseDateTime.cpp

```diff
@@ -0,0 +1,1747 @@
+            Func func{};
+            std::string func_name;
+
+            std::string literal;
+
+        public:
+            explicit Action(Func && func_, const char * func_name_) : func(std::move(func_)), func_name(func_name_) { }
+
+            explicit Action(const String & literal_) : literal(literal_) { }
+            explicit Action(String && literal_) : literal(std::move(literal_)) { }
+
+            /// For debug
+            [[maybe_unused]] String toString() const
+            {
+                if (func)
+                    return "func:" + func_name;
+                else
+                    return "literal:" + literal;
+            }
+
+            Pos perform(Pos cur, Pos end, Date & date) const
+            {
+                if (func)
+                    return func(cur, end, date);
+                else
+                {
+                    ensureSpace(cur, end, literal.size(), "required literal size not matched");
+                    if (std::string_view(cur, literal.size()) != literal)
+                        throw Exception(
+                            ErrorCodes::LOGICAL_ERROR, "Expect literal {} but {} provided", literal, std::string_view(cur, literal.size()));
+                    cur += literal.size();
+                    return cur;
+                }
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber2(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 2, "readNumber2 requires size >= 2");
+
+                res = (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                return cur;
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber3(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 3, "readNumber4 requires size >= 3");
+
+                res = (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                return cur;
+            }
+
+            template <typename T, bool check_space = true>
+            static Pos readNumber4(Pos cur, Pos end, T & res)
+            {
+                if constexpr (check_space)
+                    ensureSpace(cur, end, 4, "readNumber4 requires size >= 4");
+
+                res = (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                res = res * 10 + (*cur - '0');
+                ++cur;
+                return cur;
+            }
+
+            static void ensureSpace(Pos cur, Pos end, size_t len, const String & msg)
```

Comment:
We actually don't "ensure" space in the sense that we add some space somewhere, we just "check" space. Let's rename to `checkSpace()`.